### ISSUE TITLE
KeyError: 'array' when null values are interspersed with [] for the same dictionary key

### ISSUE THREAD

**Issue by JosXa** [koxudaxi/datamodel-code-generator#223]
**Describe the bug**

When trying to parse a JSON list of objects where items have both `null` and a list of values (or an empty list) for the same key, the library raises a `KeyError` (see example which should make it easier).

Something like this was supposedly fixed in response to #164, so it's either a regression or a new bug.

**To Reproduce**

Example JSON:

```json
{
  "my_obj": [
    {
      "items": null
    },
    {
      "items": []
    }
  ]
}
```

It doesn't matter whether the null or the [] is defined first.

The schema generated from this is:

```json
{
    "$schema": "http://json-schema.org/schema#",
    "type": "object",
    "properties": {
        "my_obj": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "items": {
                        "type": [
                            "array",
                            "null"
                        ]
                    }
                },
                "required": [
                    "items"
                ]
            }
        }
    },
    "required": [
        "my_obj"
    ]
}
```

Used `generate` call:
```
            generate(
                input_name=file.name,
                input_text=content,
                input_file_type=InputFileType.Json,
                output=out_file,
                target_python_version=PythonVersion.PY_38,
                snake_case_field=True,
            )
```

Traceback:

```
Traceback (most recent call last):

  File "C:\projects\josxabot\app\modules\josxa_private\pydantic_model_gen\directory_watcher.py", line 92, in run_conversion
    generate(
    └ <function generate at 0x00000251AF436CA0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\__init__.py", line 185, in generate
    result = parser.parse()
             │      └ <function Parser.parse at 0x00000251AF436A60>
             └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\base.py", line 351, in parse
    self.parse_raw()
    │    └ <function JsonSchemaParser.parse_raw at 0x00000251B1703430>
    └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 687, in parse_raw
    self.parse_raw_obj(obj_name, raw_obj, [])
    │    │             │         └ {'$schema': 'http://json-schema.org/schema#', 'type': 'object', 'properties': {'my_obj': {'type': 'array', 'items': {'type': ...
    │    │             └ 'Model'
    │    └ <function JsonSchemaParser.parse_raw_obj at 0x00000251B17033A0>
    └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 671, in parse_raw_obj
    self.parse_object(name, obj, path)
    │    │            │     │    └ []
    │    │            │     └ JsonSchemaObject(items=None, uniqueItem=None, type='object', format=None, pattern=None, minLength=None, maxLength=None, minim...
    │    │            └ 'Model'
    │    └ <function JsonSchemaParser.parse_object at 0x00000251B1703040>
    └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 424, in parse_object
    fields = self.parse_object_fields(obj, path)
             │    │                   │    └ []
             │    │                   └ JsonSchemaObject(items=None, uniqueItem=None, type='object', format=None, pattern=None, minLength=None, maxLength=None, minim...
             │    └ <function JsonSchemaParser.parse_object_fields at 0x00000251B16E0F70>
             └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 331, in parse_object_fields
    array_field, array_field_classes = self.parse_array_fields(
                                       │    └ <function JsonSchemaParser.parse_array_fields at 0x00000251B17030D0>
                                       └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 458, in parse_array_fields
    type=self.parse_object(
         │    └ <function JsonSchemaParser.parse_object at 0x00000251B1703040>
         └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 424, in parse_object
    fields = self.parse_object_fields(obj, path)
             │    │                   │    └ ['my_obj']
             │    │                   └ JsonSchemaObject(items=None, uniqueItem=None, type='object', format=None, pattern=None, minLength=None, maxLength=None, minim...
             │    └ <function JsonSchemaParser.parse_object_fields at 0x00000251B16E0F70>
             └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 391, in parse_object_fields
    field_types = self.get_data_type(field)
                  │    │             └ JsonSchemaObject(items=None, uniqueItem=None, type=['array', 'null'], format=None, pattern=None, minLength=None, maxLength=No...
                  │    └ <function JsonSchemaParser.get_data_type at 0x00000251B169AF70>
                  └ <datamodel_code_generator.parser.jsonschema.JsonSchemaParser object at 0x00000251AF20E6A0>

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 190, in get_data_type
    return [

  File "C:\Users\INT002327\AppData\Local\pypoetry\Cache\virtualenvs\josxabot-c3BmTbt9-py3.8\lib\site-packages\datamodel_code_generator\parser\jsonschema.py", line 192, in <listcomp>
    json_schema_data_formats[t][format_],
    │                        │  └ 'default'
    │                        └ 'array'
    └ {'integer': {'int32': <Types.int32: 2>, 'int64': <Types.int64: 3>, 'default': <Types.integer: 1>}, 'number': {'float': <Types...

KeyError: 'array'
```

**Version:**
 - OS: Windows 10
 - Python Verison: 3.8
 - datamodel-code-generator Version 0.5.34

**My super hacky workaround**

```python
def normalize_null_arrays(obj: Any):
    """ Recursively replace "null" dict values with empty lists if another item in the same list contains an empty [] """
    if isinstance(obj, dict):
        for val in obj.values():
            normalize_null_arrays(val)
        return

    if not isinstance(obj, list):
        return

    key_has_null_map = dict()
    key_has_array_map = dict()

    for i in obj:
        if not isinstance(i, dict):
            break

        for k, v in i.items():
            if v is None:
                key_has_null_map[k] = True
            if v == []:
                key_has_array_map[k] = True

    for i in obj:
        if not isinstance(i, dict):
            break

        for k, v in i.items():
            if k in key_has_array_map and k in key_has_null_map and v is None:
                i[k] = []

normalize_null_arrays(some_json_data)
```

I assume that this doesn't cover all the possible bugs that can occur, as it's a really specific example. Maybe some tests can help

**Comment by koxudaxi** [comment]
@JosXa 
Thank you for creating this issue and the workaround. 
I think the JSONSchema should be converted to below.

```python
from __future__ import annotations

from typing import Any, List, Optional
from pydantic import BaseModel

example = '''{
  "my_obj": [
    {
      "items": null
    },
    {
      "items": []
    }
  ]
}
'''

class MyObjItem(BaseModel):
    items: Optional[List[Any]]


class Model(BaseModel):
    my_obj: List[MyObjItem]

print(Model.parse_raw(example))
# my_obj=[MyObjItem(items=None), MyObjItem(items=[])]
```

How do you think about it?
if you hope it then I will impliment it. :rocket:

**Comment by JosXa** [comment]
Yeah, I suppose that's correct 👍😎

**Comment by JosXa** [comment]
On a related note (but this is just a hunch), could it be that if there are multiple models in a list and fields are set on some but None on others, the generator will use **non**-optional when it's not clear which field to use? I have to reproduce... Just leaving this here so I don't forget.

**Comment by koxudaxi** [comment]
@JosXa 
> On a related note (but this is just a hunch), could it be that if there are multiple models in a list and fields are set on some but None on others, the generator will use non-optional when in doubt? I have to reproduce... Just leaving this here so I don't forget.

Sorry, Could you show me an example?

**Comment by koxudaxi** [comment]
I have released a new version as `0.5.37`
